System/FilePath/Glob/Match.hs

```diff
@@ -122,14 +124,17 @@ match' o (OpenRange lo hi :xs) path =
 match' o again@(AnyNonPathSeparator:xs) path@(c:cs) =
    match' o xs path || (not (isPathSeparator c) && match' o again cs)
 
-match' o again@(AnyDirectory:xs) path =
-   let parts   = pathParts (dropWhile isPathSeparator path)
-       matches = any (match' o xs) parts || any (match' o again) (tail parts)
-    in if null xs && not (matchDotsImplicitly o)
-          --  **/ shouldn't match foo/.bar, so check that remaining bits don't
-          -- start with .
-          then all (not.isExtSeparator.head) (init parts) && matches
-          else matches
+match' o (AnyDirectory:xs) path =
+   if not (matchDotsImplicitly o)
```

Comment:
Flipping the then/else cases and removing this `not` would be nicer since it avoids a double negation.